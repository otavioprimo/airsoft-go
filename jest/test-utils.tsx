import React from 'react';
import {render} from '@testing-library/react-native';
import {ThemeProvider} from 'styled-components/native';
import {createStore, Reducer} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import * as Theme from '../src/theme';

interface ReduxProviders {
  state: any;
  reducer: Reducer<any, any>;
}

interface RenderProviders {
  children: React.ReactElement;
  options?: ReduxProviders;
}

const ReduxProvider = ({children, options}: any) => {
  if (!options) {
    return <>{children}</>;
  }

  const store = createStore(options.reducer, options.state);
  return <Provider store={store}>{children}</Provider>;
};

const AllTheProviders: React.FC<RenderProviders> = ({children, options}) => {
  return (
    <ThemeProvider theme={Theme.DarkTheme}>
      <ReduxProvider options={options}>{children}</ReduxProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: ReduxProviders) => {
  return render(<AllTheProviders options={options}>{ui}</AllTheProviders>);
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
