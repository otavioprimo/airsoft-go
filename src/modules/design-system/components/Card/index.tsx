import React from 'react';

import {
  Container,
  BackgroundImage,
  ContainerFooter,
  ContainerInfos,
  ContainerTop,
  Title,
  Subtitle,
  Description,
  FooterTitle,
  FooterSubTitle,
  ContainerFooterInfos,
  Icon,
  ContainerIcon,
} from './style';
import {IconProp} from '@fortawesome/fontawesome-svg-core';

interface CardProps {
  title: string;
  subtitle: string;
  description: string;
  footerTitle: string;
  image: string;
  textFirstAction?: string;
  iconFirstAction?: IconProp;
  textSecondAction?: string;
  iconSecondAction?: IconProp;
  onPress?: () => void;
  onLongPress?: () => void;
  delayLongPress?: number;
}

export const TEST_ID_BUTTON = 'button-card';

const Card: React.FC<CardProps> = ({
  image,
  title,
  subtitle,
  description,
  footerTitle,
  textFirstAction,
  iconFirstAction,
  textSecondAction,
  iconSecondAction,
  onPress,
  onLongPress,
  delayLongPress = 500,
}) => {
  const _onPress = () => {
    onPress && onPress();
  };

  const _onLongPress = () => {
    onLongPress && onLongPress();
  };

  return (
    <Container
      testID={TEST_ID_BUTTON}
      onPress={_onPress}
      onLongPress={_onLongPress}
      delayLongPress={delayLongPress}>
      <BackgroundImage
        source={{
          uri: image,
        }}
      />
      <ContainerTop>
        <ContainerInfos>
          <Subtitle>{subtitle}Ô∏è</Subtitle>
          <Title>{title}</Title>

          <Description>{description}</Description>
        </ContainerInfos>
        <ContainerFooter>
          <FooterTitle>{footerTitle}</FooterTitle>
          <ContainerFooterInfos>
            <ContainerIcon>
              <FooterSubTitle>{textFirstAction}</FooterSubTitle>
              {iconFirstAction && <Icon icon={iconFirstAction} />}
            </ContainerIcon>

            <ContainerIcon>
              <FooterSubTitle>{textSecondAction}</FooterSubTitle>
              {iconSecondAction && <Icon icon={iconSecondAction} />}
            </ContainerIcon>
          </ContainerFooterInfos>
        </ContainerFooter>
      </ContainerTop>
    </Container>
  );
};

export default Card;
