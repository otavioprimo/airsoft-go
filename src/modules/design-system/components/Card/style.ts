import styled from 'styled-components/native';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';

export const Container = styled.TouchableOpacity`
  height: 150px;
  background-color: ${({theme}) => theme.secondary};
  flex: 1;
  margin: 4px 16px;
  border-radius: 8px;
  border-color: ${({theme}) => theme.secondary};
  border-width: 0.6px;
`;
