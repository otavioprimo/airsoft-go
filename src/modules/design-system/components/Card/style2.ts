import styled from 'styled-components/native';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';

export const Container = styled.TouchableOpacity`
  height: 250px;
  background-color: ${({theme}) => theme.primary};
  flex: 0.5;
  margin: 8px;
  border-radius: 16px;
  border-color: ${({theme}) => theme.secondary};
  border-width: 0.6px;
`;

export const BackgroundImage = styled.Image.attrs({
  resizeMode: 'cover',
})`
  flex: 1;
  border-radius: 16px;
`;

export const ContainerTop = styled.View`
  position: absolute;
  height: 100%;
  width: 100%;
  border-radius: 16px;
`;

export const ContainerFooter = styled.View`
  background-color: #ccc;
  flex: 0.3;
  border-bottom-left-radius: 16px;
  border-bottom-right-radius: 16px;
  padding-left: 6px;
  padding-right: 6px;
  padding-bottom: 4px;
`;

export const ContainerInfos = styled.View`
  background-color: rgba(0, 0, 0, 0.65);
  flex: 1;
  border-top-left-radius: 16px;
  border-top-right-radius: 16px;
  padding: 8px;
`;

export const Title = styled.Text`
  color: #fff;
  font-size: 17px;
  font-weight: 800;
`;

export const Subtitle = styled.Text`
  color: #fff;
  font-size: 14px;
  margin-top: 6px;
`;

export const Description = styled.Text.attrs({
  numberOfLines: 6,
  ellipsizeMode: 'tail',
})`
  color: #fff;
  font-size: 14px;
  font-weight: 400;
  margin-top: 6px;
`;

export const ContainerFooterInfos = styled.View`
  flex: 1;
  align-items: flex-end;
  flex-direction: row;
  justify-content: space-around;
`;

export const FooterTitle = styled.Text.attrs({
  numberOfLines: 1,
})`
  margin-top: 2px;
  font-size: 16px;
  font-weight: 500;
`;

export const FooterSubTitle = styled.Text`
  color: ${({theme}) => theme.blue};
`;

export const Icon = styled(FontAwesomeIcon).attrs(({theme}) => ({
  size: 19,
  color: theme.blue,
}))`
  margin-left: 4px;
`;

export const ContainerIcon = styled.View`
  flex-direction: row;
`;
