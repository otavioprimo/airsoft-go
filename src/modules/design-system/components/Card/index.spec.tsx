import React from 'react';
import {render, act, fireEvent} from 'test/test-utils';
import {faRoute, faUsers} from '@fortawesome/free-solid-svg-icons';

import Card, {TEST_ID_BUTTON} from './index';

describe('Design System - Card', () => {
  it('Should match snapshot', () => {
    const {toJSON} = render(
      <Card
        image="https://algumacoisa.com.br/sasad.jpg"
        title="Warriors Airsoft"
        subtitle="14/06/2021"
        description={'Description 1234'}
        footerTitle="Sorocaba - SP"
        textFirstAction="250"
        iconFirstAction={faUsers}
        textSecondAction="40,4 KM"
        iconSecondAction={faRoute}
      />,
    );
    expect(toJSON()).toMatchSnapshot('DS_Card');
  });

  it('Should call onPress when click', async () => {
    const onPressProp = jest.fn();
    const {getByTestId} = render(
      <Card
        image="https://algumacoisa.com.br/sasad.jpg"
        title="Warriors Airsoft"
        subtitle="14/06/2021"
        description={'Description 1234'}
        footerTitle="Sorocaba - SP"
        textFirstAction="250"
        iconFirstAction={faUsers}
        textSecondAction="40,4 KM"
        iconSecondAction={faRoute}
        onPress={onPressProp}
      />,
    );

    await act(async () => {
      const button = getByTestId(TEST_ID_BUTTON);
      fireEvent.press(button);
      expect(onPressProp).toBeCalledTimes(1);
    });
  });

  it('Should call onLongPress when click', async () => {
    const onLongPress = jest.fn();
    const {getByTestId} = render(
      <Card
        image="https://algumacoisa.com.br/sasad.jpg"
        title="Warriors Airsoft"
        subtitle="14/06/2021"
        description={'Description 1234'}
        footerTitle="Sorocaba - SP"
        textFirstAction="250"
        iconFirstAction={faUsers}
        textSecondAction="40,4 KM"
        iconSecondAction={faRoute}
        onLongPress={onLongPress}
      />,
    );

    await act(async () => {
      const button = getByTestId(TEST_ID_BUTTON);
      fireEvent(button, 'onLongPress');
      expect(onLongPress).toBeCalledTimes(1);
    });
  });
});
