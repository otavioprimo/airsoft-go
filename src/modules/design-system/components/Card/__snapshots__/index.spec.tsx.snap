// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design System - Card Should match snapshot: DS_Card 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#222222",
      "borderColor": "rgba(230,230,230,.9)",
      "borderRadius": 16,
      "borderWidth": 0.6,
      "flexBasis": 0,
      "flexGrow": 0.5,
      "flexShrink": 1,
      "height": 250,
      "marginBottom": 8,
      "marginLeft": 8,
      "marginRight": 8,
      "marginTop": 8,
      "opacity": 1,
    }
  }
  testID="button-card"
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "https://algumacoisa.com.br/sasad.jpg",
      }
    }
    style={
      Array [
        Object {
          "borderRadius": 16,
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 16,
          "height": "100%",
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.65)",
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "paddingBottom": 8,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 8,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 14,
              "marginTop": 6,
            },
          ]
        }
      >
        14/06/2021
        Ô∏è
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 17,
              "fontWeight": "800",
            },
          ]
        }
      >
        Warriors Airsoft
      </Text>
      <Text
        ellipsizeMode="tail"
        numberOfLines={6}
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 14,
              "fontWeight": "400",
              "marginTop": 6,
            },
          ]
        }
      >
        Description 1234
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#ccc",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "flexBasis": 0,
            "flexGrow": 0.3,
            "flexShrink": 1,
            "paddingBottom": 4,
            "paddingLeft": 6,
            "paddingRight": 6,
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "marginTop": 2,
            },
          ]
        }
      >
        Sorocaba - SP
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "flexBasis": 0,
              "flexDirection": "row",
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "space-around",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#11429b",
                },
              ]
            }
          >
            250
          </Text>
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={19}
            bbWidth={19}
            fill="#11429b"
            focusable={false}
            height={19}
            meetOrSlice={0}
            minX={0}
            minY={0}
            secondaryFill={null}
            secondaryOpacity={0.4}
            size={19}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "marginLeft": 4,
                },
                Object {
                  "flex": 0,
                  "height": 19,
                  "width": 19,
                },
              ]
            }
            vbHeight={512}
            vbWidth={640}
            width={19}
          >
            <RNSVGGroup
              fill={4279321243}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGPath
                d="M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z"
                fill={4279321243}
                fillOpacity={1}
                propList={
                  Array [
                    "fill",
                    "fillOpacity",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#11429b",
                },
              ]
            }
          >
            40,4 KM
          </Text>
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={19}
            bbWidth={19}
            fill="#11429b"
            focusable={false}
            height={19}
            meetOrSlice={0}
            minX={0}
            minY={0}
            secondaryFill={null}
            secondaryOpacity={0.4}
            size={19}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "marginLeft": 4,
                },
                Object {
                  "flex": 0,
                  "height": 19,
                  "width": 19,
                },
              ]
            }
            vbHeight={512}
            vbWidth={512}
            width={19}
          >
            <RNSVGGroup
              fill={4279321243}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGPath
                d="M416 320h-96c-17.6 0-32-14.4-32-32s14.4-32 32-32h96s96-107 96-160-43-96-96-96-96 43-96 96c0 25.5 22.2 63.4 45.3 96H320c-52.9 0-96 43.1-96 96s43.1 96 96 96h96c17.6 0 32 14.4 32 32s-14.4 32-32 32H185.5c-16 24.8-33.8 47.7-47.3 64H416c52.9 0 96-43.1 96-96s-43.1-96-96-96zm0-256c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zM96 256c-53 0-96 43-96 96s96 160 96 160 96-107 96-160-43-96-96-96zm0 128c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"
                fill={4279321243}
                fillOpacity={1}
                propList={
                  Array [
                    "fill",
                    "fillOpacity",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </View>
</View>
`;
